Compatibility pytorch 2.7.0 and blackwell

johnnynunez opened the pull request:
I've fixed Kaolin so with the last commit is compatible with pytorch 2.6.0 and 2.7.0.dev
https://github.com/NVIDIAGameWorks/kaolin/pull/875

zgojcic commented:
@wilsonCernWq I know that you were running on Blackwell — can you please check this? Not sure if we should be defaulting to Cuda 12.6 as we might have problems with people that have older drivers?

johnnynunez commented:
@zgojcic done
```bash
docker build --build-arg CUDA_VERSION=12.6 -t 3dgrut:cuda126 .
```
Blackwell
```bash
docker build --build-arg CUDA_VERSION=12.8 -t 3dgrut:cuda128 .
```

zgojcic commented:
Hi @johnnynunez thanks for following up, but my comment was actually that we should keep 11.8 around (as it is on the main) because some people might have older Nvidia drivers that don't support 12.6 or 12.8 :) 

@Caenorst can you also please have a look into this as it follows the Kaoline PR? Thank you

wilsonCernWq commented:
I will test it on my setup soon, and I believe we should not change our default CUDA version to 12+ unless we have any features depending on it ---- CUDA should be backward compatible, we can merge compatibility fixes, but it is better to keep a conservative CUDA version for a wider range of users. 

Instead, I suggest we adjust this PR and merge it as an example for advanced users, who can choose to change the default settings based on their on needs. Otherwise, why we only have if conditions for CUDA 11.8, 12.6 and 12.8? This looks random to me, we should support all the possible CUDA versions, but then it would be a nightmare to maintain...

johnnynunez commented:
> I will test it on my setup soon, and I believe we should not change our default CUDA version to 12+ unless we have any features depending on it ---- CUDA should be backward compatible, we can merge compatibility fixes, but it is better to keep a conservative CUDA version for a wider range of users.
> 
> Instead, I suggest we adjust this PR and merge it as an example for advanced users, who can choose to change the default settings based on their on needs. Otherwise, why we only have if conditions for CUDA 11.8, 12.6 and 12.8? This looks random to me, we should support all the possible CUDA versions, but then it would be a nightmare to maintain...

Because the versions of cuda really depend on pytorch which you install a precompiled pytorch with x version of cuda

i revert cuda 12.6 to 11.8. So we can keep both versions (11.8-12.8(for new graphic cards)

johnnynunez commented:
done

wilsonCernWq commented:
I got errors with your current setup.

#### For 11.8, I am getting
```
Collecting git+https://github.com/NVIDIAGameWorks/kaolin (from -r requirements.txt (line 28))
  Cloning https://github.com/NVIDIAGameWorks/kaolin to /tmp/pip-req-build-2rx9hsfy
  Running command git clone --filter=blob:none --quiet https://github.com/NVIDIAGameWorks/kaolin /tmp/pip-req-build-2rx9hsfy
  Resolved https://github.com/NVIDIAGameWorks/kaolin to commit b6cf8073edbe9b5469f643d3b9cf57b0f43651db
  Running command git submodule update --init --recursive -q
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      /tmp/pip-req-build-2rx9hsfy/setup.py:12: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
        from pkg_resources import parse_version
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-req-build-2rx9hsfy/setup.py", line 53, in <module>
          raise ImportError(
      ImportError: Kaolin requires Cython >= 0.29.37. Please install it before proceeding.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```
So it seems we need to add `pip3 install cython` in the `install_env.sh`. Adding it in `requirements.txt` does not work for me.

After additionally install `cython`, I pypass this error but then hitting other errors
```
      RuntimeError:
      The detected CUDA version (12.8) mismatches the version that was used to compile
      PyTorch (11.8). Please make sure to use the same CUDA versions.
```
This seems to be you specified the conda channel incorrectly for 11.8 --- it should be `nvidia/label/cuda-11.8.0` instead of `nvidia/label/cuda-11.8`.

After fixing both issues, then I am hitting errors during training
```
  File "/media/data0/software/miniforge3/envs/3dgrut/lib/python3.11/site-packages/kornia/utils/grid.py", line 67, in create_meshgrid
    xs: Tensor = torch.linspace(0, width - 1, width, device=device, dtype=dtype)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```
I think with switching to `pip install` for torch will lead to torch 2.6.0, which does not work very well on Blackwell. I suggest we fall back to the old installation method for Cuda 11.8
```
    conda install -y cuda-toolkit -c nvidia/label/cuda-11.8.0
    conda install -y pytorch==2.1.2 torchvision==0.16.2 torchaudio==2.1.2 pytorch-cuda=11.8 "numpy<2.0" -c pytorch -c nvidia/label/cuda-11.8.0
    conda install -y cmake ninja -c nvidia/label/cuda-11.8.0
```

With all these fixes, CUDA 11.8 version starts to work for me. I am attaching a patch file for your reference.
[cuda11.patch](https://github.com/user-attachments/files/19542506/cuda11.patch)


#### For 12.8, I am getting
```
...
Channels:
 - nvidia/label/cuda-12.8.1
 - conda-forge
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: failed

InvalidSpec: The package "nvidia/label/cuda-12.8.1/linux-64::cuda-toolkit==12.8.1=0" is not available for the specified platform
```
I don't quite know why this is the case. This is my `conda info` result. I don't see anything is problematic.
```

     active environment : None
            shell level : 0
       user config file : /home/qiwu/.condarc
 populated config files : /media/data0/software/miniforge3/.condarc
                          /home/qiwu/.condarc
          conda version : 24.11.3
    conda-build version : not installed
         python version : 3.10.14.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=skylake
                          __conda=24.11.3=0
                          __cuda=12.8=0
                          __glibc=2.39=0
                          __linux=6.11.0=0
                          __unix=0=0
       base environment : /media/data0/software/miniforge3  (writable)
      conda av data dir : /media/data0/software/miniforge3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /media/data0/software/miniforge3/pkgs
                          /home/qiwu/.conda/pkgs
       envs directories : /media/data0/software/miniforge3/envs
                          /home/qiwu/.conda/envs
               platform : linux-64
             user-agent : conda/24.11.3 requests/2.31.0 CPython/3.10.14 Linux/6.11.0-19-generic ubuntu/24.04.2 glibc/2.39 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.8
                UID:GID : 1776682484:748400513
             netrc file : /home/qiwu/.netrc
           offline mode : False
```

wilsonCernWq commented:
Additionally, for Kaolin, I suggest for the stable version (CUDA 11.8), we ping kaolin to `kaolin==0.17.0`, this helps debugging.
Then only switch to the latest kaolin if necessary for Blackwell.

A possible way to implement this is remove kaolin from `requirements.txt`, and install it in the script via
```
pip install --find-links https://nvidia-kaolin.s3.us-east-2.amazonaws.com/torch-2.1.2_cu118.html kaolin==0.17.0
```

and 
```
pip install git+https://github.com/NVIDIAGameWorks/kaolin
```

And control them based on `CUDA_VERSION`.

johnnynunez commented:
> I got errors with your current setup.
> 
> #### For 11.8, I am getting
> ```
> Collecting git+https://github.com/NVIDIAGameWorks/kaolin (from -r requirements.txt (line 28))
>   Cloning https://github.com/NVIDIAGameWorks/kaolin to /tmp/pip-req-build-2rx9hsfy
>   Running command git clone --filter=blob:none --quiet https://github.com/NVIDIAGameWorks/kaolin /tmp/pip-req-build-2rx9hsfy
>   Resolved https://github.com/NVIDIAGameWorks/kaolin to commit b6cf8073edbe9b5469f643d3b9cf57b0f43651db
>   Running command git submodule update --init --recursive -q
>   Preparing metadata (setup.py) ... error
>   error: subprocess-exited-with-error
>   
>   × python setup.py egg_info did not run successfully.
>   │ exit code: 1
>   ╰─> [8 lines of output]
>       /tmp/pip-req-build-2rx9hsfy/setup.py:12: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
>         from pkg_resources import parse_version
>       Traceback (most recent call last):
>         File "<string>", line 2, in <module>
>         File "<pip-setuptools-caller>", line 34, in <module>
>         File "/tmp/pip-req-build-2rx9hsfy/setup.py", line 53, in <module>
>           raise ImportError(
>       ImportError: Kaolin requires Cython >= 0.29.37. Please install it before proceeding.
>       [end of output]
>   
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: metadata-generation-failed
> 
> × Encountered error while generating package metadata.
> ╰─> See above for output.
> 
> note: This is an issue with the package mentioned above, not pip.
> hint: See above for details.
> ```
> 
> So it seems we need to add `pip3 install cython` in the `install_env.sh`. Adding it in `requirements.txt` does not work for me.
> 
> After additionally install `cython`, I pypass this error but then hitting other errors
> 
> ```
>       RuntimeError:
>       The detected CUDA version (12.8) mismatches the version that was used to compile
>       PyTorch (11.8). Please make sure to use the same CUDA versions.
> ```
> 
> This seems to be you specified the conda channel incorrectly for 11.8 --- it should be `nvidia/label/cuda-11.8.0` instead of `nvidia/label/cuda-11.8`.
> 
> After fixing both issues, then I am hitting errors during training
> 
> ```
>   File "/media/data0/software/miniforge3/envs/3dgrut/lib/python3.11/site-packages/kornia/utils/grid.py", line 67, in create_meshgrid
>     xs: Tensor = torch.linspace(0, width - 1, width, device=device, dtype=dtype)
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> RuntimeError: CUDA error: no kernel image is available for execution on the device
> CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
> For debugging consider passing CUDA_LAUNCH_BLOCKING=1
> Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
> ```
> 
> I think with switching to `pip install` for torch will lead to torch 2.6.0, which does not work very well on Blackwell. I suggest we fall back to the old installation method for Cuda 11.8
> 
> ```
>     conda install -y cuda-toolkit -c nvidia/label/cuda-11.8.0
>     conda install -y pytorch==2.1.2 torchvision==0.16.2 torchaudio==2.1.2 pytorch-cuda=11.8 "numpy<2.0" -c pytorch -c nvidia/label/cuda-11.8.0
>     conda install -y cmake ninja -c nvidia/label/cuda-11.8.0
> ```
> 
> With all these fixes, CUDA 11.8 version starts to work for me. I am attaching a patch file for your reference. [cuda11.patch](https://github.com/user-attachments/files/19542506/cuda11.patch)
> 
> #### For 12.8, I am getting
> ```
> ...
> Channels:
>  - nvidia/label/cuda-12.8.1
>  - conda-forge
> Platform: linux-64
> Collecting package metadata (repodata.json): done
> Solving environment: failed
> 
> InvalidSpec: The package "nvidia/label/cuda-12.8.1/linux-64::cuda-toolkit==12.8.1=0" is not available for the specified platform
> ```
> 
> I don't quite know why this is the case. This is my `conda info` result. I don't see anything is problematic.
> 
> ```
> 
>      active environment : None
>             shell level : 0
>        user config file : /home/qiwu/.condarc
>  populated config files : /media/data0/software/miniforge3/.condarc
>                           /home/qiwu/.condarc
>           conda version : 24.11.3
>     conda-build version : not installed
>          python version : 3.10.14.final.0
>                  solver : libmamba (default)
>        virtual packages : __archspec=1=skylake
>                           __conda=24.11.3=0
>                           __cuda=12.8=0
>                           __glibc=2.39=0
>                           __linux=6.11.0=0
>                           __unix=0=0
>        base environment : /media/data0/software/miniforge3  (writable)
>       conda av data dir : /media/data0/software/miniforge3/etc/conda
>   conda av metadata url : None
>            channel URLs : https://conda.anaconda.org/conda-forge/linux-64
>                           https://conda.anaconda.org/conda-forge/noarch
>           package cache : /media/data0/software/miniforge3/pkgs
>                           /home/qiwu/.conda/pkgs
>        envs directories : /media/data0/software/miniforge3/envs
>                           /home/qiwu/.conda/envs
>                platform : linux-64
>              user-agent : conda/24.11.3 requests/2.31.0 CPython/3.10.14 Linux/6.11.0-19-generic ubuntu/24.04.2 glibc/2.39 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.8
>                 UID:GID : 1776682484:748400513
>              netrc file : /home/qiwu/.netrc
>            offline mode : False
> ```

about 11.8 was an error, but for cuda 12.8.1 it is on conda: https://anaconda.org/nvidia/cuda-toolkit

wilsonCernWq commented:
@johnnynunez I have incorporated your branch and made some polishing in branch `qiwu/compatibility_blackwell`. Could you give it a try to see if it works for you? I had to implement some ad-hoc fixes for torch JIT to work properly with CUDA 12.8.1 installed from conda. Simple training of 3DGRT and 3DGUT works for me now with torch nightly and CUDA 12.8 on RTX 5090 (compiled natively to CC 12.0).

**NOTE**: I changed the script behavior to parse CUDA_VERSION as an environment variable instead of a positional argument. This makes the script more easier to use in my opinion as the option will not be confused with the `WITH_GCC11` flag.

**NOTE**: I temporarily disabled the installation of kaolin. Could you incorporate @orperel's suggestion on your side? You seem to have the right knowledge for that 🙏

johnnynunez commented:
> @johnnynunez I have incorporated your branch and made some polishing in branch `qiwu/compatibility_blackwell`. Could you give it a try to see if it works for you? I had to implement some ad-hoc fixes for torch JIT to work properly with CUDA 12.8.1 installed from conda. Simple training of 3DGRT and 3DGUT works for me now with torch nightly and CUDA 12.8 on RTX 5090 (compiled natively to CC 12.0).
> 
> **NOTE**: I changed the script behavior to parse CUDA_VERSION as an environment variable instead of a positional argument. This makes the script more easier to use in my opinion as the option will not be confused with the `WITH_GCC11` flag.
> 
> **NOTE**: I temporarily disabled the installation of kaolin. Could you incorporate @orperel's suggestion on your side? You seem to have the right knowledge for that 🙏

It works for me.
Anyways, i can compile kaolin with blackwell support. Maybe when pytorch 2.7.0 comes out (4/23/25) you can upload a kaolin compiled with pytorch 2.7.0 and blackwell support

wilsonCernWq commented on install_env.sh:
```diff
@@ -82,9 +83,24 @@ if [ "$WITH_GCC11" = true ]; then
+    pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu128
+    conda install -y cmake ninja -c nvidia/label/cuda-12.8.1
+    pip3 install --force-reinstall "numpy<2"
+    IGNORE_TORCH_VER=1 pip3 install git+https://github.com/NVIDIAGameWorks/kaolin
```
I should have resolved this.

johnnynunez commented:
I have running 3dgrut on GH200 with cuda 12.8... cheers! https://pypi.jetson-ai-lab.dev/sbsa/cu128
